theopensystemslab/planx-new
# feat: Send component makes a single request to create Hasura Scheduled Events

Currently, if a service submits to multiple destinations, it will have multiple Send components in its' graph. These are causing noticible user delays between Pay and the Confirmation page; the `/uniform` request especially takes awhile to complete.

In an effort to remedy this, we're refactoring as follows:
- One Send component per flow that allows an editor to select multiple destinations
- Send will now always make the same request to the `/create-send-events/:sessionId` API endpoint
- The API endpoint will create one Hasura Scheduled Event per selected destination - these will show up in the Hasura console under Events > [One-off Scheduled Events](https://hasura.io/docs/latest/scheduled-triggers/create-one-off-scheduled-event/#:~:text=One%2Doff%20scheduled%20events%20are,after%20a%20user%20signs%20up.)
- Hasura Events will handle making the requests to `/bops` & `/uniform` including retries

Trade-offs:
- With two Send components in the graph, we were able to Send the application id returned by BOPS to Uniform. We lose the ability to do that with independently scheduled events. A [common application number that we fetch upfront](https://trello.com/c/7QJ3iuF8/1046-send-planning-case-reference-from-bops-to-pay-awaiting-bops-api-change) would solve this
- We lose the ability to display an error message in the public-facing Send if this local authority isn't configured for the selected destination (which was just a nice-to-have piece of feedback for editors). Now, an event would still be created, but fail with a status 400 and the custom error message

Clean-up:
- Only show the `<SaveResumeButton />` on Cards that have a visible "Continue" button
- Add a `send/` directory to the API project and move BOPS & Uniform methods out of server.js for maintainability (rename `send.js` > `send/uniform.js` etc)
- Rename Editor method `getParams` > `getBOPSParams` for clarity

Future: 
- Add a more explicit "submission to <destination>" analytics event so we don't need to rely on retaining data in audit tables
- Validate on "publish" that a flow has one Send component (easy to manually manage in content right now)

**To test:**
- Go through an LDC flow https://1127.planx.pizza/lambeth/apply-for-a-lawful-development-certificate/preview?analytics=false (I've updated this Send compoent & published)
- Pay using [test details](https://docs.payments.service.gov.uk/testing_govuk_pay/), confirm in network tab that Send events are successfully created, see Confirmation page
- Check that the events were processed in Hasura Console https://hasura.1127.planx.pizza/console/events/one-off-scheduled-events/processed
- Check that `bops_applications` & `uniform_applications` tables have receipt records from successful event runs
